common/lib/xmodule/xmodule/modulestore/__init__.py

```diff
@@ -139,6 +140,20 @@ def unnest(self):
         """
         self._active_count -= 1
 
+    def add_publish_item(self, location):
+        """
+        Adds an item to the set of things that have been published
+        """
+        self._publish_items.add(location.version_agnostic().for_branch(None))
```

Comment:
Performance penalty for doing a full course reindex - really depends upon the course - for the demo course in a sandbox it's less than 1 second; for 6.002x (which is the largest that I've seen) it is about 13 seconds.

However, Cale has suggested that there's some low hanging fruit in walking the course as well; I could see a case for optimizing this instead of worrying about all the differences. Re-indexing is only a problem for the time that it takes - and really it's only the crawling of the course-tree that is the time sink herein (the actual index is updated asynchronously anyway).

Using `subtree_edited_on` could certainly help us avoid adding unnecessary stuff to the same index as before, but that's not where we'd be time-bound within the celery task.

Perhaps I should step back this PR to just the first commit, and then worry about adding extra granularity only if it turns out that we need to do so?